### PR TITLE
fix(views): adds dehydration calls to ng-repeat removed views.

### LINKED ISSUE #416: view dehydration
Views are not automatically de-hydrated when removed from a viewport (they are hydrated when created).

`NgRepeat` removes views when:
- items are being removed from the list of items,
- items are being moved in the lists of items.

When items are being moved, there is no need to dehydrate the view as the view will still be present in the end DOM. However when views are being removed, they should be de-hydrated.

`ViewPort.clear()` remove all the attached view. They should probably also be dehydrated there.

### PR DESCRIPTION
**Pull request by rkirov**
Closes #416

### CODE REVIEW

File: modules/directives/src/ng_repeat.js [hunk 1]
```diff
@@ -60,9 +60,13 @@ export class NgRepeat extends OnChange {
     for (var i = tuples.length - 1; i >= 0; i--) {
       var tuple = tuples[i];
       var view = viewPort.remove(tuple.record.previousIndex);
+      // separate moved views from removed views.
       if (isPresent(tuple.record.currentIndex)) {
         tuple.view = view;
         ListWrapper.push(movedTuples, tuple);
+      } else {
+        // TODO(rado): add a test for dehydration when onDestroy is available. 
+        view.dehydrate();
```
Comment (vsavkin) [inline modules/directives/src/ng_repeat.js]: I thought hydration/dehydration is supposed to be hidden from the developer. If so, maybe `viewPort` should support `moves` explicitly, so dehydration can done there?

Comment (rkirov) [inline modules/directives/src/ng_repeat.js]: Created a new API method in viewPort called `detach` that is meant to used for view moves (i.e. move the nodes, keep the directives). Added an extra test using onDestroy. Please take another look.

File: modules/core/src/compiler/viewport.js [hunk 2]
```diff
@@ -89,7 +89,18 @@ export class ViewPort {
...
+    var view = this.detach(atIndex);
+    view.dehydrate();
+    // view is intentionally not returned to the client.
+  }
+
+  /**
+   * The method can be used together with insert to implement a view move, i.e.
+   * moving the dom nodes while the directives in the view stay intact.
+   */
+  detach(atIndex=-1): View {
```
Comment (vicb) [inline modules/core/src/compiler/viewport.js]: may be rename `var removedView` inside the method

File: modules/directives/test/ng_repeat_spec.js [hunk 3]
```diff
@@ -230,3 +257,27 @@ class TestComponent {
     this.items = [1, 2];
   }
 }
+
+@Decorator({
+  selector: 'lifetime-dec',
+  lifecycle: ["onDestroy"]
```
Comment (vicb) [inline modules/directives/test/ng_repeat_spec.js]: use `onDestroy` (var as opposed to string)

File: modules/core/test/compiler/viewport_spec.js [hunk 4]
```diff
@@ -92,13 +91,31 @@ export function main() {
...
         expect(viewPort.length).toBe(2);
       });
 
+      it('should detach the last view by default', () => {
+        viewPort.insert(customViewWithOneNode);
+
+        var detachedView = viewPort.detach();
+
+        expect(detachedView).toBe(customViewWithOneNode);
+        expect(viewPort.length).toBe(1);
```
Comment (vicb) [inline modules/core/test/compiler/viewport_spec.js]: it would be easier to understand if you `expect(viewPort.length).toBe(2);` before the view is detached - same (3) with the following test

File: modules/core/test/compiler/viewport_spec.js [hunk 5]
```diff
@@ -81,10 +81,9 @@ export function main() {
       it('should remove the last view by default', () => {
```
Comment (vicb) [inline modules/core/test/compiler/viewport_spec.js]: What about checking that `customViewWithOneNode` start dehydrated, is hydrated on insert & dehydrated on remove.
May be it should be an other test.

File: modules/directives/test/ng_repeat_spec.js [hunk 6]
```diff
@@ -190,6 +201,21 @@ export function main() {
     });
   });
 
+  it('should destroy all directives in removed views', (done) => {
```
Comment (vsavkin) [inline modules/directives/test/ng_repeat_spec.js]: Is it really NgRepeat's concern? NgRepeat just removes views. It does not directly destroy directives. It is just a part of the "remove" operation. 

Or is it a way to test that we actually remove views and not just detach them?

**Comment by vsavkin** [comment]
LGTM

File: modules/core/src/compiler/viewport.js (hunk 2 above)
Comment (rkirov) [inline modules/core/src/compiler/viewport.js]: done.

File: modules/directives/test/ng_repeat_spec.js (hunk 3 above)
Comment (rkirov) [inline modules/directives/test/ng_repeat_spec.js]: onDestroy in not declared `const`. I am leaving this for another PR, there are issues with `const` and the dart transpiler. TODO added.

File: modules/core/test/compiler/viewport_spec.js (hunk 4 above)
Comment (rkirov) [inline modules/core/test/compiler/viewport_spec.js]: done.

File: modules/core/test/compiler/viewport_spec.js (hunk 5 above)
Comment (rkirov) [inline modules/core/test/compiler/viewport_spec.js]: actually insert and detach do not change the hydration state of the view, that's their whole point - to enable move for clients. Added tests for them.
